### PR TITLE
Classifying the ISIC 2020 kaggle challenge data set with the use of a Siamese Network - 47044232

### PR DESCRIPTION
**Pull request by DeadmanJ**
Hi team,

This pull request contains my solution to the Siamese classification problem in task 6.
The results I get from classifying the test set after training sits between 79% and 80% on average.
I have added the following files:
- `README.md` Documents what I have done. Please go here to have full explanation of what was done and and detailed results.
- `dataset.py` Handles the class storing the dataset and the function to get the data loaders.
- `modules.py` Contains the code for both the Siamese network model and the binary classifier.
- `utils.py` Contains the code for some helper functions used throughout the project. Mainly used to provide readability in code.
- `config.py` The go to place for changing any hyperparameters or changing where data is loaded and stored to.
- `train.py` Contains the code for training, testing and validating both the Siamese network and the binary classifier.
- `predict.py` Contains the code for performing inference on a sample set the data using two saved models.
- `./images/` The images used by the README file.
- `.gitignore` To ignore files generated that shouldn't go into the repository.

Thanks ahead of time,
Joshua Deadman
47044232

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **20**

Solves problem: The solution is well-suited to the problem (**5**). 
_The training manifold is effectively separated, and the classifier shows strong performance, as reflected by the accuracy and reported confusion matrix. 
Triplet loss is utilized, which is suitable for the problem.
Class imbalance is addressed through undersampling of benign cases._

Implementation functions as intended  (**2**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:
1. Well understaning of the medical images challanges and the triplet loss.
2. Future suggestions are well crafted and might improve the performance.
3. the specificity (76.07%) and sensitivity (88.03%) are almost balanced which means the model is able to differentiate between the two classess efficiently

_Using an algorithmic approach to undersample the majority class is indeed recommended over random sampling, especially in the context of medical imaging_ to maintain informative samples, improve generalization, and avoide loss of critical information.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work!
